actors/ppo_atari_pz.py

```diff
@@ -0,0 +1,1085 @@
+
+
+class PPOActor:
+    """PPO actor"""
+
+    def __init__(self, _cfg):
+        self._dtype = torch.float
+
+    def get_actor_classes(self):
+        """Get actor"""
+        return [PLAYER_ACTOR_CLASS]
+
+    async def impl(self, actor_session):
+        # Start a session
+        actor_session.start()
+        config = actor_session.config
+
+        # Get observation and action space
+        environment_specs = EnvironmentSpecs.deserialize(config.environment_specs)
+        observation_space = environment_specs.get_observation_space()
+        action_space = environment_specs.get_action_space(seed=config.seed)
+
+        # Get model
+        model, _, version_info = await actor_session.model_registry.retrieve_version(
+            PPOModel, config.model_id, config.model_version
+        )
+        # log.info(f"Actor - retreved model number: {version_info['version_number']}")
+        obs_shape = model.input_shape[::-1]
+
+        async for event in actor_session.all_events():
+            if event.observation and event.type == cogment.EventType.ACTIVE:
+                if (
+                    event.observation.observation.HasField("current_player")
+                    and event.observation.observation.current_player != actor_session.name
+                ):
+                    # Not the turn of the agent
+                    actor_session.do_action(action_space.serialize(action_space.create()))
+                    continue
+                obs = observation_space.deserialize(event.observation.observation)
+                obs_tensor = torch.tensor(obs.flat_value, dtype=self._dtype).reshape(obs_shape)
+                obs_tensor = torch.unsqueeze(obs_tensor.permute((2, 0, 1)), dim=0).to(torch.device(model.device))
+
+                # Get action from policy network
+                with torch.no_grad():
+                    dist = model.network.get_action(obs_tensor)
+                    action = dist.sample().cpu().numpy()[0]
+
+                # Send action to environment
+                if isinstance(action_space.gym_space, Discrete):
+                    action_value = action_space.create(value=action)
+                    actor_session.do_action(action_space.serialize(action_value))
+                else:
+                    # TODO: to be tested with continuous actions
+                    action_value = action_space.create(value=action)
+                    actor_session.do_action(action_space.serialize(action_value))
+
+
+class BasePPOTraining(ABC):
+    """Abstract class for training PPO agent"""
+
+    default_cfg = {
+        "seed": 0,
+        "num_epochs": 10,
+        "num_iter": 500,
+        "epoch_num_trials": 1,
+        "num_parallel_trials": 1,
+        "discount_factor": 0.99,
+        "entropy_loss_coef": 0.05,
+        "value_loss_coef": 0.5,
+        "action_loss_coef": 1.0,
+        "clipping_coef": 0.1,
+        "learning_rate": 3e-4,
+        "batch_size": 64,
+        "num_steps": 2048,
```

Comment:
not really clear what's that supposed to be.

Reply:
I was changed to -1 corresponding to the end of the trial. A comment is added to explain this number